File: 2020/16_TicketTranslation/test_rule.py
# ======================================================================
# Ticket Translation
#   Advent of Code 2020 Day 16 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         t e s t _ r u l e . p y
# ======================================================================
"Test Rule class for Advent of Code 2020 day 16, Ticket Translation"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import rule

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE_TEXT = "row: 6-11 or 33-44"

# ======================================================================
#                                                               TestRule
# ======================================================================


class TestRule(unittest.TestCase):  # pylint: disable=R0904
    "Test Rule object"

    def test_empty_init(self):
        "Test the default Rule creation"

        # 1. Create default Rule object
        myobj = rule.Rule()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, None)
        self.assertEqual(myobj.name, None)
        self.assertEqual(myobj.ranges, [])

    def test_text_init(self):
        "Test the Rule object creation from text"

        # 1. Create Rule object from text
        myobj = rule.Rule(EXAMPLE_TEXT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, EXAMPLE_TEXT)
        self.assertEqual(myobj.name, 'row')
        self.assertEqual(len(myobj.ranges), 2)
        self.assertEqual(myobj.ranges[0][0], 6)
        self.assertEqual(myobj.ranges[0][1], 11)
        self.assertEqual(myobj.ranges[1][0], 33)
        self.assertEqual(myobj.ranges[1][1], 44)

        # 3. Test methods
        self.assertEqual(myobj.is_valid(1), False)
        self.assertEqual(myobj.is_valid(6), True)
        self.assertEqual(myobj.is_valid(7), True)
        self.assertEqual(myobj.is_valid(11), True)
        self.assertEqual(myobj.is_valid(12), False)
        self.assertEqual(myobj.is_valid(40), True)
        self.assertEqual(myobj.is_valid(45), False)
        self.assertEqual(myobj.is_valid(50), False)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      t e s t _ r u l e . p y                   end
# ======================================================================

File: 2020/16_TicketTranslation/test_rule.lua
-- ======================================================================
-- Ticket Translation
--   Advent of Code 2020 Day 16 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ r u l e . l u a
-- ======================================================================
-- Test Rule for Advent of Code 2020 day 16, Ticket Translation

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local rule = require('rule')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = "row: 6-11 or 33-44"

-- ======================================================================
--                                                               TestRule
-- ======================================================================

function test_empty_init()
  -- Test the default Rule creation

  -- 1. Create default Rule object
  local myobj = rule:Rule()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(myobj.name, '???')
  luaunit.assertEquals(#myobj.numbers, 0)

end

function test_text_init()
  -- Test the Rule object creation from text

  -- 1. Create Tickets object from text
  local myobj = rule:Rule({text=EXAMPLE_TEXT})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.text, EXAMPLE_TEXT)
  luaunit.assertEquals(myobj.name, 'row')
  luaunit.assertEquals(#myobj.numbers, 4)
  luaunit.assertEquals(myobj.numbers[1], 6)
  luaunit.assertEquals(myobj.numbers[2], 11)
  luaunit.assertEquals(myobj.numbers[3], 33)
  luaunit.assertEquals(myobj.numbers[4], 44)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:is_valid(0), false)
  luaunit.assertEquals(myobj:is_valid(5), false)
  luaunit.assertEquals(myobj:is_valid(6), true)
  luaunit.assertEquals(myobj:is_valid(8), true)
  luaunit.assertEquals(myobj:is_valid(11), true)
  luaunit.assertEquals(myobj:is_valid(12), false)
  luaunit.assertEquals(myobj:is_valid(15), false)
  luaunit.assertEquals(myobj:is_valid(32), false)
  luaunit.assertEquals(myobj:is_valid(33), true)
  luaunit.assertEquals(myobj:is_valid(38), true)
  luaunit.assertEquals(myobj:is_valid(44), true)
  luaunit.assertEquals(myobj:is_valid(45), false)
  luaunit.assertEquals(myobj:is_valid(55), false)
  
  luaunit.assertEquals(myobj:are_all_valid({}), true)
  luaunit.assertEquals(myobj:are_all_valid({5}), false)
  luaunit.assertEquals(myobj:are_all_valid({5, 8}), false)
  luaunit.assertEquals(myobj:are_all_valid({8, 5}), false)
  luaunit.assertEquals(myobj:are_all_valid({8, 38}), true)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                     t e s t _ r u l e . p y                    end
-- ======================================================================
